### PR TITLE
mm: Epoch reporting

### PR DESCRIPTION
**Pull request by martonp**
```
This diff updates the market makers to generate a report of their
activities during each epoch. If the rates of each placement is unable
to be determined, this is reported as a `PreOrderProblem`, otherwise
an `OrderReport` is generated for the orders placed on each side of the
market containing information such as lots the bot is required to place,
the current number of lots standing, the number of lots booked, the
balances required and used for each placement, etc. This information is
all displayed on the UI.

A bug is also fixed in this diff. Previously, on a call to
`core.MultiTrade`, it was possible for some of the orders to be placed,
and some of them to result in an error. In this case, an error was returned
from `core.MultiTrade`, and the market makers would assume that none
of the trades were placed. Now a `core.MultiTradeResult` is returned for
each of the requested orders in the `MultiTrade` call containing either
the order or an error.
```

Screenshot 2024-10-29 at 3 59 37 PM

Screenshot 2024-10-29 at 4 00 00 PM
Screenshot 2024-10-29 at 4 00 20 PM

### CODE REVIEW

File: client/mm/exchange_adaptor.go [hunk 1]
```diff
@@ -3303,6 +3366,146 @@ func (u *unifiedExchangeAdaptor) Book() (buys, sells []*core.MiniOrder, _ error)
...
+			problems.CEXTradeErr = nil
+		} else {
+			problems.CEXTradeErr = newStampedError(err)
+		}
+	})
+}
+
+// checkBotHealth returns true if the bot is healthy and can continue trading.
+func (u *unifiedExchangeAdaptor) checkBotHealth() bool {
+	user := u.clientCore.User()
```
Comment (buck54321) [inline client/mm/exchange_adaptor.go:3441]: Pulling the entire `core.User` is way too much. Please find or create a more efficient way to get the little bit of data you need from core.

File: client/mm/mm.go [hunk 2]
```diff
@@ -205,12 +208,149 @@ type CEXStatus struct {
...
+}
+
+func (bp *BotProblems) copy() *BotProblems {
+	copy := *bp
+
+	if bp.WalletNotSynced != nil {
+		copy.WalletNotSynced = make(map[uint32]bool, len(bp.WalletNotSynced))
+		for k, v := range bp.WalletNotSynced {
+			copy.WalletNotSynced[k] = v
+		}
```
Comment (buck54321) [inline client/mm/mm.go:299]: `utils.CopyMap`

File: client/mm/mm_basic.go [hunk 3]
```diff
@@ -411,6 +435,52 @@ func (m *basicMarketMaker) ordersToPlace(basisPrice, feeAdj uint64) (buyOrders,
...
+			problems.DeterminePlacementsErr = determinePlacementsErr
+		}
+
+		problems.DEXBalanceDeficiencies = dexDefs
+	})
+}
+
+// clearAllOrders returns slices of order placements with zero rate and lots.
+// When these are passed to multiTrade, all orders will be cancelled.
+func clearAllOrders(numBuyOrders, numSellOrders int) (buyOrders, sellOrders []*multiTradePlacement) {
```
Comment (buck54321) [inline client/mm/mm_basic.go:466]: Woudn't it be easier to create a method that just cancels all of the orders? What's the advantage of plumbing the through `mutliTrade`?

File: client/mm/mm_arb_market_maker.go [hunk 4]
```diff
@@ -258,6 +256,7 @@ func (a *arbMarketMaker) ordersToPlace() (buys, sells []*multiTradePlacement) {
 			}
 
 			if !filled {
+				cexTooShallow[map[bool]string{true: "sell", false: "buy"}[!sellOnDEX]] = true
```
Comment (buck54321) [inline client/mm/mm_arb_market_maker.go:259]: `sellStr`?

File: client/mm/exchange_adaptor.go [hunk 5]
```diff
@@ -1066,6 +1070,44 @@ func (u *unifiedExchangeAdaptor) multiTrade(
 
 	orderInfos := make([]*dexOrderInfo, 0, len(requiredPlacements))
 
+	dexDeficiencies, cexDeficiencies = func() (dexDeficiencies, cexDeficiencies map[uint32]uint64) {
```
Comment (buck54321) [inline client/mm/exchange_adaptor.go:1073]: Why is this a function?

**Review by buck54321 — CHANGES_REQUESTED** [review]

**Review by JoeGruffins — APPROVED** [review]

**Comment by JoeGruffins** [comment]
Unsure if related to these changes but will see a lot of errors running on simnet with a bot:
```
2024-06-27 06:11:02.064 [ERR] CORE: notify: |ERROR| (order) Swap reporting error - Error notifying DEX of swap for match 75738e143d8220928c55c0599ac65c3b2c21df15cfff1da538ea52911b64aea7: error sending 'init' message: timed out waiting for "init" response (timeout) - Order: 4a527c9e2e9f4985f79e90d20823fecb60e1b75954850fe11b6371a1fbf60f02
2024-06-27 06:11:02.064 [ERR] CORE[wss://127.0.0.1:17273/ws]: No handler found for response: {"type":2,"id":196,"payload":{"result":null,"error":{"code":24,"message":"failed to find contract coin f3714c2c324210f2c9c28e81b7136c1ecf59d2bd32d693bc5f37a998322d32f9:0"}},"sig":""}
```
The one below this may be fixed in #2843
```
2024-06-27 06:11:45.180 [ERR] MM[MM-127.0.0.1:17273-60-0]: Error getting tx 68434de5cbaa0e27d5204512db08c45e3487ddfcc077fe5d4f8bbd396f09c1cf00000000: tx database not initialized
```
```
2024-06-27 06:13:45.063 [ERR] CORE: failed to send trade request: new order request with DEX server 127.0.0.1:17273 market eth_btc failed: rpc error: error code 53: order quantity exceeds user limit
```

Can be fixed later.

File: client/mm/exchange_adaptor.go [hunk 6]
```diff
@@ -3466,6 +3598,167 @@ func (u *unifiedExchangeAdaptor) Book() (buys, sells []*core.MiniOrder, _ error)
...
+
+type cexProblemType uint16
+
+const (
+	cexTradeProblem cexProblemType = iota
+	cexDepositProblem
+	cexWithdrawProblem
+)
+
+func (u *unifiedExchangeAdaptor) updateCEXProblems(typ cexProblemType, assetID uint32, err error) {
```
Comment (dev-warrior777) [inline client/mm/exchange_adaptor.go:3659]: Could this func be simplified?

**Review by dev-warrior777 — APPROVED** [review]
I just looked for general code issues since I am not as familiar as I should be with this area of the code.
UI looks great btw.

**Review by buck54321 — APPROVED** [review]
This is stellar work. I could nit-pick at stuff and I have some followup, but I'm going to merge this as is. It's going in the patch.

File: client/mm/exchange_adaptor.go (hunk 6 above)
Comment (buck54321) [reply client/mm/exchange_adaptor.go:3659]: I don't disagree. The locking scheme seems like overkill. We'll follow up.

**Pull request merged**